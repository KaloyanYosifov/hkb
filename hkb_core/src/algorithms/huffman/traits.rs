use std::collections::VecDeque;

use crate::data_structures::NodeRef;

use super::{HuffmanNode, HuffmanValue};

type HuffmanNodeRef = NodeRef<HuffmanValue>;

fn traverse_node<T, F: Fn(&String, &HuffmanNode) -> Option<T>>(
    node: &HuffmanNode,
    callback: F,
) -> Option<T> {
    let mut queue: VecDeque<(String, HuffmanNodeRef)> = VecDeque::with_capacity(32);

    if let Some(node) = node.get_left() {
        queue.push_front(("0".to_string(), node))
    }

    if let Some(node) = node.get_right() {
        queue.push_front(("1".to_string(), node))
    }

    while !queue.is_empty() {
        let (binary, node) = queue.pop_front().unwrap();
        let borrowed = node.borrow();
        let return_val = callback(&binary, &borrowed);

        if return_val.is_some() {
            return return_val;
        }

        if let Some(node) = borrowed.get_left() {
            queue.push_front((format!("{}0", binary), node))
        }

        if let Some(node) = borrowed.get_right() {
            queue.push_front((format!("{}1", binary), node))
        }
    }

    None
}

pub trait HuffmanNodeTraverser {
    fn to_binary(&self, c: char) -> Option<String>;
    fn from_binary(&self, binary_encoded: String) -> Option<char>;
}

// TODO: Find out if the solutions below are ok for us
// or if we should find a better faster solution
impl HuffmanNodeTraverser for HuffmanNode {
    fn to_binary(&self, c: char) -> Option<String> {
        traverse_node(self, |binary, node| {
            if Some(c) == node.val.char {
                return Some(binary.clone());
            } else {
                None
            }
        })
    }

    fn from_binary(&self, binary_encoded: String) -> Option<char> {
        traverse_node(self, move |binary, node| {
            if binary_encoded.as_str() == binary.as_str() && node.val.char.is_some() {
                return node.val.char;
            } else {
                None
            }
        })
    }
}

#[cfg(test)]
mod tests {
    use std::{collections::HashSet, hash};

    use insta::assert_snapshot;

    use crate::algorithms::{HuffmanEncoder, HuffmanNode, HuffmanNodeTraverser, HuffmanValue};

    #[test]
    fn it_can_convert_char_to_binary() {
        let node = HuffmanNode::new(
            HuffmanValue {
                char: None,
                occurance: 2,
            },
            Some(HuffmanValue {
                char: Some('a'),
                occurance: 1,
            }),
            Some(HuffmanValue {
                char: Some('b'),
                occurance: 1,
            }),
        );

        assert_eq!("0", node.to_binary('a').unwrap().as_str());
        assert_eq!("1", node.to_binary('b').unwrap().as_str());
        assert!(matches!(node.to_binary('c'), None));
    }

    #[test]
    fn it_can_convert_from_binary_to_char() {
        let node = HuffmanNode::new(
            HuffmanValue {
                char: None,
                occurance: 2,
            },
            Some(HuffmanValue {
                char: Some('a'),
                occurance: 1,
            }),
            Some(HuffmanValue {
                char: Some('b'),
                occurance: 1,
            }),
        );

        assert_eq!('a', node.from_binary("0".to_string()).unwrap());
        assert_eq!('b', node.from_binary("1".to_string()).unwrap());
        assert!(matches!(node.from_binary("01".to_string()), None));
    }

    #[test]
    fn it_can_convert_char_to_binary_snap() {
        let text = "Hello there magnificent mothertrucker";
        let node = HuffmanEncoder::compress(text);
        let mut output = String::with_capacity(1024);
        let mut chars = text
            .chars()
            .collect::<HashSet<_>>()
            .into_iter()
            .collect::<Vec<_>>();

        chars.sort();

        for c in chars {
            output
                .push_str(format!("Character {} = {} \n", c, node.to_binary(c).unwrap()).as_str());
        }

        assert_snapshot!(output);
    }

    #[test]
    fn it_can_convert_binary_to_char() {
        let text = "Hello there magnificent mothertrucker";
        let node = HuffmanEncoder::compress(text);

        let expected: Vec<(char, &str)> = vec![
            ('H', "01010"),
            ('c', "1011"),
            ('k', "01111"),
            ('n', "11110"),
            ('t', "100"),
            ('l', "11111"),
            ('o', "0110"),
            (' ', "1110"),
            ('r', "001"),
            ('e', "110"),
            ('i', "1010"),
            ('m', "0100"),
            ('u', "01110"),
            ('a', "01011"),
            ('f', "00011"),
            ('g', "00010"),
            ('h', "0000"),
        ];

        for (expected_char, binary_encoded) in expected.into_iter() {
            let char = node.from_binary(binary_encoded.to_string()).unwrap();

            assert_eq!(expected_char, char);
        }
    }
}
