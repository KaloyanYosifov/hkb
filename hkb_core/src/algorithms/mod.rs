pub use huffman::*;

mod huffman;
