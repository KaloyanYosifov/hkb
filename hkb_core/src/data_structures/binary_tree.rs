use std::{cell::RefCell, rc::Rc};

trait Constraints: PartialEq + Eq + PartialOrd + Ord {}

impl<T: PartialEq + Eq + PartialOrd + Ord> Constraints for T {}

pub type NodeRef<T> = Rc<RefCell<Node<T>>>;

#[derive(Ord, Eq, Debug)]
pub struct Node<T: PartialEq + Eq + PartialOrd + Ord> {
    pub val: T,
    left: Option<NodeRef<T>>,
    right: Option<NodeRef<T>>,
}

impl<T: Constraints> PartialEq for Node<T> {
    fn eq(&self, other: &Self) -> bool {
        self.val == other.val
    }
}

impl<T: Constraints> PartialOrd for Node<T> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        let val = {
            if self.val > other.val {
                std::cmp::Ordering::Greater
            } else if self.val < other.val {
                std::cmp::Ordering::Less
            } else {
                std::cmp::Ordering::Equal
            }
        };

        Some(val)
    }
}

impl<T: Constraints> Node<T> {
    pub fn new(val: T, left_val: Option<T>, right_val: Option<T>) -> Self {
        let left = {
            if let Some(v) = left_val {
                Some(Rc::new(RefCell::new(Node::with_value(v))))
            } else {
                None
            }
        };
        let right = {
            if let Some(v) = right_val {
                Some(Rc::new(RefCell::new(Node::with_value(v))))
            } else {
                None
            }
        };

        Self { val, left, right }
    }

    pub fn with_value(val: T) -> Self {
        Self {
            val,
            left: None,
            right: None,
        }
    }

    pub fn with_nodes(val: T, left: Node<T>, right: Node<T>) -> Self {
        Self {
            val,
            left: Some(Rc::new(RefCell::new(left))),
            right: Some(Rc::new(RefCell::new(right))),
        }
    }

    pub fn get_left<'a>(&self) -> Option<NodeRef<T>> {
        if let Some(node) = self.left.as_ref() {
            Some(node.clone())
        } else {
            None
        }
    }

    pub fn get_right<'a>(&self) -> Option<NodeRef<T>> {
        if let Some(node) = self.right.as_ref() {
            Some(node.clone())
        } else {
            None
        }
    }

    pub fn height(&self) -> usize {
        self.height_recursive(self)
    }

    fn height_recursive(&self, node: &Node<T>) -> usize {
        match (node.get_left(), node.get_right()) {
            (Some(left), Some(right)) => {
                let left_height = self.height_recursive(&left.borrow());
                let right_height = self.height_recursive(&right.borrow());
                1 + std::cmp::max(left_height, right_height)
            }
            (Some(left), None) => 1 + self.height_recursive(&left.borrow()),
            (None, Some(right)) => 1 + self.height_recursive(&right.borrow()),
            (None, None) => 1,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Node;

    #[test]
    fn it_can_create_a_node_with_value() {
        let node = Node::with_value(3);

        assert_eq!(3, node.val);
        assert!(matches!(node.get_left(), None));
        assert!(matches!(node.get_right(), None));
    }

    #[test]
    fn it_can_create_a_node_with_left_right_moved() {
        let node = Node::new(3, Some(5), Some(7));

        assert_eq!(3, node.val);
        assert_eq!(5, node.get_left().unwrap().borrow().val);
        assert_eq!(7, node.get_right().unwrap().borrow().val);
    }

    #[test]
    fn it_can_create_with_nodes_moved_directly() {
        let left_node = Node::with_value(5);
        let right_node = Node::with_value(7);
        let node = Node::with_nodes(3, left_node, right_node);

        assert_eq!(3, node.val);
        assert_eq!(5, node.get_left().unwrap().borrow().val);
        assert_eq!(7, node.get_right().unwrap().borrow().val);
    }

    #[test]
    fn it_can_return_the_height_of_the_node_when_it_is_one() {
        let node = Node::with_value(3);

        assert_eq!(1, node.height());
    }

    #[test]
    fn it_can_return_the_height_of_the_node() {
        let left_node = Node::with_value(5);
        let right_node = Node::with_value(7);
        let node = Node::with_nodes(3, left_node, right_node);

        assert_eq!(2, node.height());
    }
}
