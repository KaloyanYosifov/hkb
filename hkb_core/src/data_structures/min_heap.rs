/// Min heap
/// 1. Always the most minimum value is on top
/// 2. We have a trait which returns an integer value
/// 3. The trait will be dynamic and boxed
///
/// Implementation
/// 1. Insert:
///     - Put the value in the end
///     - Bubble up
/// 2. Pop
///     - Keep the value we are going to return
///     - Swap with the bottom most element
///     - remove it from the last element
///     - Start from root and heapify to fix the top minimum value

// TODO:
// Maybe we can do traits support here and a user can handle any internal obj they want to keep
// Something like MinHeapValue trait and the user can implement a struct with this trait and the
// struct can have { node: Some(node), node_val: i64, somehting: None }

pub struct MinHeap {
    values: Vec<i64>,
}

impl MinHeap {
    pub fn new() -> Self {
        Self::with_capacity(32)
    }

    pub fn with_capacity(capacity: usize) -> Self {
        Self {
            values: Vec::with_capacity(capacity),
        }
    }
}

impl MinHeap {
    pub fn insert(&mut self, value: i64) {
        self.values.push(value);

        self.heapify_bottom_up();
    }

    pub fn pop(&mut self) -> Option<i64> {
        if self.values.len() == 0 {
            None
        } else if self.values.len() == 1 {
            self.values.pop()
        } else {
            self.swap(0, self.values.len() - 1);

            let value = self.values.pop();

            self.heapify_top_down();

            value
        }
    }

    fn heapify_bottom_up(&mut self) {
        let mut index = self.values.len() - 1;
        let mut parent = index / 2;

        while self.values[parent] > self.values[index] {
            self.swap(index, parent);
            index = parent;
            parent = index / 2;
        }
    }

    fn heapify_top_down(&mut self) {
        let mut index = 0;

        loop {
            let current_val = self.values[index];
            let left = (index * 2) + 1;
            let right = (index * 2) + 2;

            if left < self.values.len() && right < self.values.len() {
                let left_val = self.values[left];
                let right_val = self.values[right];

                if left_val < right_val && left_val < current_val {
                    self.swap(left, index);
                    index = left;
                } else if right_val < left_val && right_val < current_val {
                    self.swap(right, index);
                    index = right;
                } else {
                    break;
                }
            } else if left < self.values.len() {
                let left_val = self.values[left];

                if left_val < current_val {
                    self.swap(left, index);
                    index = left;
                } else {
                    break;
                }
            } else if right < self.values.len() {
                let right_val = self.values[right];

                if right_val < current_val {
                    self.swap(right, index);
                    index = right;
                } else {
                    break;
                }
            } else {
                break;
            }
        }
    }

    fn swap(&mut self, index1: usize, index2: usize) {
        let temp = self.values[index1];
        self.values[index1] = self.values[index2];
        self.values[index2] = temp;
    }
}

#[cfg(test)]
mod tests {
    use super::MinHeap;

    // TODO: add property testing

    fn create_min_heap() -> MinHeap {
        let mut min_heap = MinHeap::new();
        let mut values: [usize; 10] = core::array::from_fn(|i| i + 1);

        values.reverse();

        for i in values {
            min_heap.insert(i as i64);
        }

        min_heap
    }

    #[test]
    fn it_can_insert_a_value() {
        let mut min_heap = MinHeap::new();

        min_heap.insert(5);

        assert_eq!(5, min_heap.values[0]);
    }

    #[test]
    fn it_puts_the_min_value_at_top() {
        let mut min_heap = MinHeap::new();
        let mut values: [usize; 10] = core::array::from_fn(|i| i + 1);

        values.reverse();

        for i in values {
            min_heap.insert(i as i64);
            assert_eq!(i as i64, min_heap.values[0]);
        }

        assert_eq!(1, min_heap.values[0]);
    }

    #[test]
    fn it_does_not_put_new_high_value_to_the_top() {
        let mut min_heap = create_min_heap();

        min_heap.insert(300);

        assert_eq!(1, min_heap.values[0]);
    }

    #[test]
    fn it_can_get_top_value_from_min_heap() {
        let mut min_heap = create_min_heap();

        let val = min_heap.pop();

        assert_eq!(1, val.unwrap());
        assert_eq!(2, min_heap.values[0]);
    }
}
