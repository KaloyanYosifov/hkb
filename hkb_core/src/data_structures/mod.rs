pub mod binary_tree;
pub mod min_heap;
