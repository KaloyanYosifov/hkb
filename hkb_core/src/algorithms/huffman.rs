use std::collections::HashMap;
pub use traits::*;

mod traits;

use crate::data_structures::{MinHeap, Node};

pub type HuffmanNode = Node<HuffmanValue>;

#[derive(Eq, Ord, Clone, Copy, Debug)]
pub struct HuffmanValue {
    pub char: Option<char>,
    pub occurance: u64,
}

impl PartialEq for HuffmanValue {
    fn eq(&self, other: &Self) -> bool {
        self.char == other.char
    }
}

impl PartialOrd for HuffmanValue {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        let val = {
            if self.occurance > other.occurance {
                std::cmp::Ordering::Greater
            } else if self.occurance < other.occurance {
                std::cmp::Ordering::Less
            } else {
                std::cmp::Ordering::Equal
            }
        };

        Some(val)
    }
}

pub struct HuffmanEncoder;

impl HuffmanEncoder {
    pub fn compress(text: &str) -> HuffmanNode {
        let mut huffman_values: Vec<HuffmanValue> = Vec::with_capacity(text.len() / 2);
        let mut occurance_map: HashMap<char, usize> = HashMap::with_capacity(text.len() / 2);

        for c in text.chars().into_iter() {
            let index = occurance_map
                .entry(c)
                .or_insert_with(|| huffman_values.len() + 1);
            let actual_index = *index - 1;

            if actual_index >= huffman_values.len() {
                huffman_values.push(HuffmanValue {
                    char: Some(c),
                    occurance: 0,
                });
            }

            huffman_values[actual_index].occurance += 1;
        }

        // TODO: Check if we can do this more efficiently without using an intermediate vector
        // and if we can just directly push to the min heap and store a pointer to a map to update
        // the struct whenever we need to
        let mut priority_queue: MinHeap<HuffmanNode> = MinHeap::with_capacity(text.len() / 2);
        for value in huffman_values.into_iter() {
            priority_queue.insert(HuffmanNode::with_value(value));
        }

        while priority_queue.size() > 1 {
            let el1 = priority_queue.pop().unwrap();
            let el2 = priority_queue.pop().unwrap();

            let occurance = el1.val.occurance + el2.val.occurance;
            let node = HuffmanNode::with_nodes(
                HuffmanValue {
                    char: None,
                    occurance,
                },
                el1,
                el2,
            );

            priority_queue.insert(node);
        }

        priority_queue.pop().unwrap()
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::HuffmanEncoder;

    #[test]
    fn it_generates_correct_snapshot() {
        let text = "Hello there magnificent mothertrucker";

        assert_debug_snapshot!(HuffmanEncoder::compress(text));
    }
}
