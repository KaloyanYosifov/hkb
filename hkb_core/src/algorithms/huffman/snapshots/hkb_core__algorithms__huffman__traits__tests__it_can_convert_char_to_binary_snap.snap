---
source: hkb_core/src/algorithms/huffman/traits.rs
expression: output
---
Character   = 1110 
Character H = 01010 
Character a = 01011 
Character c = 1011 
Character e = 110 
Character f = 00011 
Character g = 00010 
Character h = 0000 
Character i = 1010 
Character k = 01111 
Character l = 11111 
Character m = 0100 
Character n = 11110 
Character o = 0110 
Character r = 001 
Character t = 100 
Character u = 01110
